### DISCUSSION TITLE
Any dynamic route renders 4x times(!!!) in the browser

### DISCUSSION THREAD

**Discussion by michal-mad** [vercel/next.js#40749] [Help] [answered]
Issue closely related to the following bug report: [https://github.com/vercel/next.js/issues/12010](url) but this time I noticed that *any* dynamic page is rendered 4 times (!) which is quite alarming IMO.

`const Page = () => {
  console.log('rendered')
  return <div />
}
export default Page`

Result:
`undefined
undefined
rendered
rendered`

I found out that it behaves like that under those circumstance:

- `react` and `react-dom` version 18.0.0 and up (with version 17.x it renders as expected - twice).
- by using `next dev`. With `next start` it behaves as expected.

Tested on Edge and Chrome.
Looks to me as a bug, but wanted to ask here first if thats "normal" by any means..

**Comment by icyJoseph** [comment, marked as answer]
Well, React 18 introduced changes to Strict Mode, in so that it, during development mode, https://reactjs.org/docs/strict-mode.html#ensuring-reusable-state:

1. Mounts the component
2. Un-mounts immediately
3. Mounts again

Now, combine that with the query property and how hydration works, as described here, https://nextjs.org/docs/advanced-features/automatic-static-optimization#how-it-works:

> During prerendering, the router's query object will be empty since we do not have query information to provide during this phase. After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.
> 
> The cases where the query will be updated after hydration triggering another render are:
> 
> The page is a [dynamic-route](https://nextjs.org/docs/routing/dynamic-routes).
> The page has query values in the URL.
> [Rewrites](https://nextjs.org/docs/api-reference/next.config.js/rewrites) are configured in your next.config.js since these can have parameters that may need to be parsed and provided in the query.

You can use console counts to get a better view, :

```tsx
// pages/foo/[bar].tsx
import type { NextPage } from "next";
import { useRouter } from "next/router";
import { useEffect } from "react";

const Bar: NextPage = () => {
  console.count("Bar");
  const { isReady, query } = useRouter();
  console.log({ isReady, query });

  useEffect(() => {
    console.count("Effect runs");
    return () => {
      console.count("Effect clean-up");
    };
  }, []);

  return <div>Dynamic Route</div>;
};

export default Bar;
```

And with Strict Mode on this logs:

```
Bar: 1
{isReady: false, query: {…}}
Bar: 2
{isReady: false, query: {…}}
Effect runs: 1
Effect clean-up: 1
Effect runs: 2
Bar: 3
{isReady: true, query: {…}}
Bar: 4
{isReady: true, query: {…}}
```

With Strict Mode off:

```
Bar: 1
{isReady: false, query: {…}}
Effect runs: 1
Bar: 2
{isReady: true, query: {…}}
```

To turn on/off `StrictMode` in Next.js:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: false,
};

module.exports = nextConfig;
```

> I think there's also something going on with the wrappers Next.js puts around our application code, because I'd expect triple rendering, not four, but perhaps I am missing something here.

Now we take it a step further. You don't actually need a dynamic route, just pass a query to the url like: `http://localhost:3000/?foo=bar`, and:

```tsx
// pages/index.tsx
const Home = () => {
  console.count("home"); // counts four times if there's a query
  return <div>Home</div>;
};

export default Home;
```

And it logs:

```
home: 1
home: 2
home: 3
home: 4
```

Remove the query and we are back at double count. 

This starts to delve into implementation details about how Next.js tries to make the queries available to the route object, and the hydration process. Do remember that there's a few other Next.js specifics wrapping your application code.

If you wanna learn how that works, the `Performance` timings are a good start, but I'd say that, since these are implementation details, and the various re-renderings do not happen in production mode, it is enough with understanding what's Next.js doing, what's React's part, and learn to implement clean up functions.

**Reply by zillBoy** [reply]
Thanks turning reactStrictMode to `false` and using `getServerSideProps` fixed it for me. Now rendering only 1 time 👍 

```
const nextConfig = {
  reactStrictMode: false
 }
```

```
export const getServerSideProps = async (appContext) => {
  const pokemonData = appContext.query;
  return { props: { pokemonData: pokemonData || null } };
};
```
